src/irmin-pack/inode_intf.ml

```diff
@@ -89,7 +89,14 @@ module type Internal = sig
 
     val of_raw : (key -> Raw.t option) -> Raw.t -> t
     val to_raw : t -> Raw.t
-    val save : add:(hash -> Raw.t -> unit) -> mem:(key -> bool) -> t -> unit
+
+    val save :
```

Comment:
I'm not sure what these argument to save do. :(

Reply:
`mem` tests the presence of the key in the underlying storage, `save _ raw` pushes [raw] into the underlying storage

Reply:
Called only once here: https://github.com/mirage/irmin/blob/craig%40pack-direct-pointers/src/irmin-pack/inode.ml#L1729-L1733.

This `Internal` module type is a fairly random collection of functions that was at some point split out into its own functor for testing elsewhere. We should trim it down in a refactoring pass.

Reply:
Let me check/ ask some more questions:
- add: takes a hash and a raw and adds it to the underlying storage (without adding to the index)
- index: takes a hash and returns (Some key) if the corresponding key exists in the index (and keys in the index imply there must be an object in the store)
- mem: checks whether a key already exists (in the index? or the underlying pack store?)
?

Reply:
> add: ... without adding to the index

Whether or not the value is added to the index is a function of the indexing strategy, so it _may_ be (and adding a value to a store with `add` or `unsafe_add` is currently the only way to trigger it to index something).

> index: ... keys in the index imply there must be an object in the store

This is true for `irmin-pack` (which is all the inodes are used for), but not of all `index` implementations for Irmin backend stores, as clarified in [the documentation](https://github.com/mirage/irmin/blob/craig%40pack-direct-pointers/src/irmin/indexable_intf.ml#L35-L50). (Most of them implement `index` as `fun _ hash -> Some hash`.)

> mem: ... in the index? or the underlying pack store?

This depends on the internal state of the key. If we have only the hash, we must check the index; for `Direct` keys we ensure:
- the key references a valid offset;
- the key is not corrupt (by reading the actual hash at that offset from the pack file).

https://github.com/mirage/irmin/blob/craig%40pack-direct-pointers/src/irmin-pack/pack_store.ml#L230-L271